spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/mapping/IdGeneratingBeforeSaveCallback.java

```diff
@@ -45,26 +45,47 @@ public Object onBeforeSave(Object aggregate, MutableAggregateChange<Object> aggr
         Assert.notNull(aggregate, "The aggregate cannot be null at this point");
 
         RelationalPersistentEntity<?> persistentEntity = relationalMappingContext.getPersistentEntity(aggregate.getClass());
+
+        if (persistentEntity.getIdProperty() == null) {
+            return aggregate;
+        }
+
+        // we're doing INSERT and ID property value is not set explicitly by client
+        if (persistentEntity.isNew(aggregate) && !idIsProvidedByClient(aggregate, persistentEntity)) {
```

Comment:
Naming isn't ideal. What is a client?

`hasIdentifierValue` should be just fine.